### ISSUE TITLE
Overhaul gruel

### ISSUE THREAD

**Issue by cratuki** [solent-eng/solent#97]
The whole gruel system is a mess. The functionality is incomplete, the layout of the module is a mess. I suspect the TDD approach is flawed and creates complication also.

The first piece of work is to simply work out a sensible approach to taming it.

**Comment by cratuki** [comment]
Currently, the solent test suite fails when I run it under Windows, but works when I run it under linux. The place where it fails is in the tests for gruel. Even though I want to do an overhaul of gruel, this is interesting. There is no good reason that the tests should behave differently across platforms. Hence, good first step is to fix this unit tests and understand the failure.

**Comment by cratuki** [comment]
OK. The error is due to slightly different behaviour between windows vs other platforms on how TCP timeouts behave. It is not significant.

Next step: build a mental picture of how you would consolidate the nearcast for gruel_server into spin_gruel_server. When this code was written, I did not have a strong opinion about how to lay out nearcasts. Below, I will highlight some rules I would like to bring it into line with.

// Rule: nearcast definition and cogs live in the same file

In a nearcast application, the launcher file should contain the following,
- Bootstrap (create an engine, create an orb, populate the orb, get it running)
- Nearcast definition (interface script, typically captured in I_NEARCAST)
- The cogs that belong to that nearcast definition.

// Rule: Business logic does not live in cogs

When you are attaching business logic to a nearcast, we do it like this,
- We create rails to house the functionality.
- We use cogs as a thin gearing between the rail's API and the nearcast.

.

The server looks like a natural attack point for this work. Once we have that well-structured, it should be straightforward to them go back and refactor the client to the same rules.

Something else I notice is that I went to some effort to consolidate the protocol definitions in a single file. This could be seen as having the same intent as something like Protocol Buffers (much more primitive). This is probably not necessary. At the beginning of the year, Gruel looked intimidating and complex. Now, as far as protocols go, it looks trivial. Since there is no complexity to house, there is no need to try and force the protocol objects into a central definition.

Instead: have server module, and then a separate client module.

**Comment by cratuki** [comment]
There is no clear pressing need for gruel, and I do not want the distraction of this issue or non-working code in the codebase. So I have moved the code out. It continues to live in tree history, and I have made a copy of the files in my private file store at `/home/cturner/saga/20171221.nearlab.cturner.aa.solent.gruel.retirement`. There is nothing stopping us from bringing it out of retirement later if we deem it useful.

Gut-feel.

* it will/would be easier to rewrite from scratch than to recover in its current form

* but the current implementation would be informative on the design of such work.
